src/pages/patientView/clinicalInformation/PatientViewPageStore.ts

```diff
@@ -158,31 +161,42 @@ export class PatientViewPageStore
         ],
         invoke: () => {
-            const sampleToEntrezGeneIds = _.reduce(this.mutationData.result, (map:{ [s:string]:Set<number> }, next:Mutation) => {
-                const sampleId = next.sampleId;
-                map[sampleId] = map[sampleId] || new Set();
-                map[sampleId].add(next.entrezGeneId);
-                return map;
-            }, {});
-
-            return new Promise((resolve, reject) => {
+            return Promise.resolve(function(sampleToEntrezGeneIds:{ [sampleId:string]:Set<number> }) {
+                // Reference to new immutable map
+                // See which we need to fetch, and set "pending" for those data
+                const toQuery:{ [sampleId:string]:number[]} = {};
+                for (const sampleId in sampleToEntrezGeneIds) {
+                    if (sampleToEntrezGeneIds.hasOwnProperty(sampleId)) {
+                        this.mrnaExprRankData[sampleId] = this.mrnaExprRankData[sampleId] || {};
+                        for (const entrezGeneId of sampleToEntrezGeneIds[sampleId]) {
+                            if (!this.mrnaExprRankData[sampleId].hasOwnProperty(entrezGeneId)) {
+                                toQuery[sampleId] = toQuery[sampleId] || [];
+                                toQuery[sampleId].push(entrezGeneId);
+                                this.mrnaExprRankData[sampleId][entrezGeneId] = { status:"pending" };
+                            }
+                        }
+                    }
+                }
+                // Fetch that data
                 const mrnaPercentiles: MrnaPercentile[] = [];
-                const fetchAllMrnaPercentilesPromise = Promise.all(Object.keys(sampleToEntrezGeneIds).map((sampleId:string) =>
+                const fetchAllMrnaPercentilesPromise = Promise.all(Object.keys(toQuery).map((sampleId:string) =>
                     new Promise((sampleResolve, sampleReject) => {
-                        const entrezGeneIds = sampleToEntrezGeneIds[sampleId];
-                        if (typeof entrezGeneIds === "undefined" || entrezGeneIds.size === 0 || this.mrnaRankGeneticProfileId.result === null) {
+                        const entrezGeneIds = toQuery[sampleId];
+                        if (this.mrnaRankGeneticProfileId.result === null) {
                             sampleResolve();
                         } else {
                             const fetchMrnaPercentilesPromise = internalClient.fetchMrnaPercentileUsingPOST({
                                 geneticProfileId:this.mrnaRankGeneticProfileId.result,
-                                sampleId:sampleId,
-                                entrezGeneIds: Array.from(entrezGeneIds)
+                                sampleId,
+                                entrezGeneIds
                             });
                             fetchMrnaPercentilesPromise.then((d) => {
                                 mrnaPercentiles.push.apply(mrnaPercentiles, d);

```

Comment:
I was having some sort of issue with setting `mrnaPercentiles = mrnaPercentiles.concat(d)`, which is why I opted to use `.push` to do it in-place. Good call on the spread operator.